private/model/api/example.go

```diff
@@ -0,0 +1,342 @@
+}
+
+func generateExample(ex Example) string {
+	buf := bytes.NewBuffer(nil)
+	buf.WriteString(fmt.Sprintf("input := &%s.%s {\n", ex.API.PackageName(), ex.Operation.InputRef.Shape.ShapeName))
+	if ex.Operation.HasInput() {
+		buf.WriteString(buildShape(&ex.Operation.InputRef, ex.Input, false))
+	}
+	buf.WriteString("}\n")
+	buf.WriteString(fmt.Sprintf("\nresult, err := svc.%s(input)\n", ex.OperationName))
+	return buf.String()
+}
+
+// generateTypes will generate no types for default examples, but customizations may
+// require their own defined types.
+func generateTypes(ex Example) string {
+	return ""
+}
+
+// buildShape will recursively build the referenced shape based on the json object
+// provided.
+func buildShape(ref *ShapeRef, shapes map[string]interface{}, isMap bool) string {
+	order := make([]string, len(shapes))
+	for k := range shapes {
+		order = append(order, k)
+	}
+	sort.Strings(order)
+
+	ret := ""
+	for _, name := range order {
+		shape := shapes[name]
+
+		// If the shape isn't a map, we want to export the value, since every field
+		// defined in our shapes are exported.
+		if len(name) > 0 && !isMap && strings.ToLower(name[0:1]) == name[0:1] {
+			name = strings.Title(name)
+		}
+
+		memName := name
+		t := ""
+		if isMap {
+			memName = fmt.Sprintf("%q", memName)
+		} else if ref != nil {
+		}
+
+		switch v := shape.(type) {
+		case bool:
+			ret += fmt.Sprintf("%s: aws.Bool(%t),\n", memName, v)
+		case int:
+			if ref.Shape.MemberRefs[name].Shape.Type == "timestamp" {
+				ret += parseTimeString(ref, memName, fmt.Sprintf("%d", v))
+			} else {
+				ret += fmt.Sprintf("%s: aws.Int64(%d),\n", memName, v)
+			}
+		case string:
+			if ref != nil && ref.Shape.MemberRefs[name] != nil && ref.Shape.MemberRefs[name].Shape.Type == "timestamp" {
+				ret += parseTimeString(ref, memName, fmt.Sprintf("%s", v))
+			} else if ref != nil && ref.Shape.MemberRefs[name] != nil && ref.Shape.MemberRefs[name].Shape.Type == "blob" {
+				if (ref.Shape.MemberRefs[name].Streaming || ref.Shape.MemberRefs[name].Shape.Streaming) && ref.Shape.Payload == name {
+					ret += fmt.Sprintf("%s: aws.ReadSeekCloser(bytes.NewBuffer([]byte(%q))),\n", memName, v)
+				} else {
+					ret += fmt.Sprintf("%s: []byte(%q),\n", memName, v)
+				}
+			} else {
+				ret += fmt.Sprintf("%s: aws.String(%q),\n", memName, v)
+			}
+		case map[string]interface{}:
+			if ref == nil {
+				ret += fmt.Sprintf(`%s: {
+				%s
+			},
+			`, memName, buildShape(nil, v, isMap))
+			} else {
+				if isMap {
+					t = ref.Shape.Type
+				} else {
+					t = ref.Shape.MemberRefs[name].Shape.Type
+				}
+				switch t {
+				case "structure":
+					shapeName := ""
+					if ref.Shape.MemberRefs[name] != nil {
+						shapeName = ref.Shape.MemberRefs[name].Shape.ShapeName
+					} else {
+						shapeName = ref.Shape.ShapeName
+					}
+					ret += fmt.Sprintf(`%s: &%s.%s{
+				%s
+			},
+			`, memName, ref.API.PackageName(), shapeName, buildShape(ref.Shape.MemberRefs[name], v, false))
```

Comment:
This is a really good example of these template strings should be a const or template and not defined inline.